src/Microsoft.Restier.Core/ApiBase.cs

```diff
@@ -85,7 +85,10 @@ protected ApiConfiguration Configuration
                     apiType =>
                     {
                         IServiceCollection services = new ServiceCollection()
-                            .CutoffPrevious<IQueryExecutor>(DefaultQueryExecutor.Instance);
+                            .CutoffPrevious<IQueryExecutor>(DefaultQueryExecutor.Instance)
+                            .AddSingleton<ApiConfigurationExtensions.PropertyBag>()
+                            .AddScoped<ApiContextExtensions.PropertyBag>()
+                            .AddScoped<InvocationContextExtensions.PropertyBag>();
```

Comment:
1 scoped PropertyBag service would be enough since you already make InvocationContext a dedicated scope.